upcoming: [DI-23788] - Changes for resource selection limitation

venkymano-akamai opened the pull request:
## Description 📝

Highlight the Pull Request's context and intentions.

## Changes  🔄

List any change(s) relevant to the reviewer.

- ...
- ...

## Target release date 🗓️

Please specify a release date (and environment, if applicable) to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷

**Include a screenshot or screen recording of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| 📷 | 📷 |

## How to test 🧪

### Prerequisites

(How to setup test environment)

- ...
- ...

### Reproduction steps

(How to reproduce the issue, if applicable)

- [ ] ...
- [ ] ...

### Verification steps

(How to verify changes)

- [ ] ...
- [ ] ...

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [ ] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [ ] All unit tests are passing
- [ ] TypeScript compilation succeeded without errors
- [ ] Code passes all linting rules

</details>

---

## Commit message and pull request title format standards

> **Note**: Remove this section before opening the pull request
**Make sure your PR title and commit message on squash and merge are as shown below**

`<commit type>: [JIRA-ticket-number] - <description>`

**Commit Types:**

- `feat`: New feature for the user (not a part of the code, or ci, ...).
- `fix`: Bugfix for the user (not a fix to build something, ...).
- `change`: Modifying an existing visual UI instance. Such as a component or a feature.
- `refactor`: Restructuring existing code without changing its external behavior or visual UI. Typically to improve readability, maintainability, and performance.
- `test`: New tests or changes to existing tests. Does not change the production code.
- `upcoming`: A new feature that is in progress, not visible to users yet, and usually behind a feature flag.

**Example:** `feat: [M3-1234] - Allow user to view their login history`

---

venkymano-akamai commented:
Hi Team, This is a work in progress, will open once it is ready

abailly-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Resources/CloudPulseModifyAlertResources.tsx:
```diff
@@ -21,6 +23,18 @@ export const CloudPulseModifyAlertResources = React.memo(
+
+      return flags.aclpAlertServiceTypeConfig?.find(
+        ({ serviceType }) => serviceType === serviceTypeWatcher
+      )?.maxResourceSelectionCount;
```
I am not crazy about this, mostly because while typed, data from Launch Darkly can be altered on the fly and break things. The service can also break and return undefined, so destructuring is dangerous.

abailly-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsResources/AlertsResourcesNotice.tsx:
```diff
@@ -40,38 +38,46 @@ export const AlertsResourcesNotice = React.memo(
           >
             Select All
-          </LinkButton>
+          </Button>
```
use one button and ternaries for the label & data-ids

venkymano-akamai replied:
I think Having a button per operation is more readable than a ternary operator in this case.

abailly-akamai replied:
no

venkymano-akamai replied:
@abailly-akamai , done this

abailly-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsResources/AlertsResources.tsx:
```diff
@@ -401,6 +414,15 @@ export const AlertResources = React.memo((props: AlertResourcesProp) => {
           </Grid>
         )}
+        <AlertsNoticeMessage text={errorText} variant="error" />
+        <AlertsNoticeMessage
+          text={
+            maxSelectionCount !== undefined
+              ? `You can select up to ${maxSelectionCount} resources.`
+              : undefined
+          }
+          variant="warning"
+        />
```
hmm you would get an empty notice..

```suggestion
{maxSelectionCount !== undefined && (
        <AlertsNoticeMessage
          text=`You can select up to ${maxSelectionCount} resources.`
          variant="warning"
        />
      )
```

venkymano-akamai replied:
@abailly-akamai , this is not needed as AlertsNoticeMessage component handles this, if the text is undefined it returns a empty component rather a empty notice

abailly-akamai replied:
Please clean up nonetheless

venkymano-akamai replied:
@abailly-akamai , done this

abailly-akamai reviewed (commented):
Thanks for working on this,. You have a lot of type assertions in this PR, which is a code smell. Please address them before further review:

`as ObjectSchema<EditAlertDefinitionPayload>`
`as ObjectSchema<CreateAlertDefinitionForm>`

will review the UI once this (and other comments) have been addressed

jaalah-akamai commented on packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts:
```diff
@@ -251,3 +275,36 @@ export const processMetricCriteria = (
+          CreateAlertDefinitionForm | EditAlertDefinitionPayload
+        >
+      );
+};
```
After spending quite some time with your code, I think we can eliminate all the type assertions. It started getting messy when trying to use the same validation for both edit/create even though they were similar. It's much cleaner (type-wise) just to keep them separate but share the internals.

```
export const schemaWithEntityIdsValidation = <
  T extends CreateAlertDefinitionForm | EditAlertDefinitionPayload
>(
  baseSchema: ObjectSchema<T>,
  serviceTypeObj: null | string,
  aclpAlertServiceTypeConfig: AclpAlertServiceTypeConfig[]
): ObjectSchema<T> => {

  if (!serviceTypeObj || !aclpAlertServiceTypeConfig.length) {
    return baseSchema;
  }

  const config = aclpAlertServiceTypeConfig.find(
    ({ serviceType: configServiceType }) => configServiceType === serviceTypeObj
  );

  const maxSelectionCount = config?.maxResourceSelectionCount;

  if (maxSelectionCount === undefined) {
    return baseSchema;
  }

  return baseSchema.test({
    exclusive: true,
    message: `The overall number of resources assigned to an alert can't exceed ${maxSelectionCount}.`,
    name: 'entity-ids-max',
    test(value) {
      // We need to check if entity_ids exists and is an array
      if (
        !value ||
        !('entity_ids' in value) ||
        !Array.isArray(value.entity_ids)
      ) {
        return true;
      }

      return value.entity_ids.length <= maxSelectionCount;
    },
  });
};
```
Instead of modifying the schema directly using `concat` we're just using Yup's `.test()` method.

Then we can break this out into both and edit and create schema:
```
interface SchemaValidationProps<T extends object> {
  aclpAlertServiceTypeConfig: AclpAlertServiceTypeConfig[];
  baseSchema: ObjectSchema<T>;
  serviceTypeObj: null | string;
}

// Create a generic function that works with both
// schema types to avoid eslint no-identical-functions
const getSchema = <T extends object>(
  props: SchemaValidationProps<T>
): ObjectSchema<T> => {
  const { aclpAlertServiceTypeConfig, baseSchema, serviceTypeObj } = props;
  return schemaWithEntityIdsValidation(
    baseSchema,
    serviceTypeObj,
    aclpAlertServiceTypeConfig
  );
};

export const getCreateAlertSchema = (
  props: SchemaValidationProps<CreateAlertDefinitionForm>
): ObjectSchema<CreateAlertDefinitionForm> => {
  return getSchema<CreateAlertDefinitionForm>(props);
};

export const getEditAlertSchema = (
  props: SchemaValidationProps<EditAlertDefinitionPayload>
): ObjectSchema<EditAlertDefinitionPayload> => {
  return getSchema<EditAlertDefinitionPayload>(props);
};

jaalah-akamai replied:
If you planned a different approach in the follow-up ticket you created, feel free to disregard - but I think this would work. The test updates are simple with this approach too.

venkymano-akamai replied:
@jaalah-akamai , since we are eliminating type assertions in the upcoming PR, I have already did a similar approach by splitting the function into two for create and edit and just concat without need of type assertion. I think we can do this work in the upcoming PR itself

Code references of upcoming work where type assertions are removed and similar thing is achieved

https://github.com/venkymano-akamai/manager/blob/resource_limitation_with_type_fix/packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts#L272

https://github.com/venkymano-akamai/manager/blob/resource_limitation_with_type_fix/packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts#L286

Moving common part to a shared function
https://github.com/venkymano-akamai/manager/blob/resource_limitation_with_type_fix/packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts#L300

Type assertions removed
https://github.com/venkymano-akamai/manager/blob/resource_limitation_with_type_fix/packages/manager/src/features/CloudPulse/Alerts/CreateAlert/CreateAlertDefinition.tsx#L74
https://github.com/venkymano-akamai/manager/blob/resource_limitation_with_type_fix/packages/manager/src/features/CloudPulse/Alerts/EditAlert/EditAlertDefinition.tsx#L54

We will take this in upcoming task as we are already on it. May be we can use this test instead of concat there as well.

venkymano-akamai commented:
@jaalah-akamai , Thanks for all the opinions, I have addressed all the comments and replied for a comment regarding schemas with upcoming work, trying to achieve something similar. Lets take that up as a upcoming work.

venkymano-akamai commented:
Merging, all jobs and checks passed, enough approvals.

Type assertion fix is pointed and will be raised in next PR.
